ttnn/cpp/ttnn/operations/ccl/common/host/ccl_command_stream_builders.cpp

```diff
@@ -0,0 +1,181 @@
+    const int64_t larger_slice_size = smaller_slice_size + !evenly_divisible;
+
+    auto compute_slice_dim_size =
+        [larger_slice_size, smaller_slice_size, num_larger_slices_total](int64_t slice_index) {
+            bool is_larger_slice = slice_index < num_larger_slices_total;
+            return is_larger_slice ? larger_slice_size : smaller_slice_size;
+        };
+
+    auto compute_slice_offset = [num_larger_slices_total, larger_slice_size, smaller_slice_size](int64_t slice_index) {
+        int64_t num_larger_slices = std::min(slice_index, num_larger_slices_total);
+        int64_t num_smaller_slices = std::min(slice_index - num_larger_slices, 0L);
+        return num_larger_slices * larger_slice_size + (slice_index - num_larger_slices) * smaller_slice_size;
+    };
+
+    auto compute_slice_size_and_offset = [compute_slice_dim_size,
+                                          compute_slice_offset](size_t slice_index) -> std::pair<size_t, size_t> {
+        return {compute_slice_dim_size(slice_index), compute_slice_offset(slice_index)};
+    };
+    auto result = std::vector<std::pair<size_t, size_t>>{};
+    result.reserve(num_slices);
+    for (size_t i = 0; i < num_slices; i++) {
+        result.push_back(compute_slice_size_and_offset(i));
+    }
+    return std::vector<std::pair<size_t, size_t>>(result.begin(), result.end());
+}
+
+// // Outer vector = per worker command stream, inner vector = commands
+std::vector<std::vector<ttnn::ccl::v2::TensorSlice>> split_tensor_slices_across_workers_page_aligned(
+    size_t num_workers, std::vector<ttnn::ccl::v2::TensorSlice> const& tensor_slices) {
+    TT_FATAL(tensor_slices.size() > 0, "Number of slices must be greater than 0");
+    // not split up across workers yet
+
+    auto worker_slices_streams = std::vector<std::vector<ttnn::ccl::v2::TensorSlice>>(num_workers);
+    std::ranges::for_each(
+        worker_slices_streams, [&tensor_slices](auto& worker_slices) { worker_slices.reserve(tensor_slices.size()); });
+
+    for (auto const& tensor_slice : tensor_slices) {
+        auto const worker_slices = split_tensor_slice_across_workers_wrapped_page_aligned(tensor_slice, num_workers);
+        TT_FATAL(
+            worker_slices.size() == num_workers,
+            "Expected {} worker slices for tensor slice but got {}",
+            num_workers,
+            worker_slices.size());
+        for (size_t i = 0; i < num_workers; i++) {
+            worker_slices_streams[i].push_back(worker_slices[i]);
+        }
+    }
+    for (size_t i = 0; i < num_workers; i++) {
+        TT_FATAL(
+            worker_slices_streams[i].size() == tensor_slices.size(),
+            "Mismatch in tensor slices. Expected {} but got {}",
+            tensor_slices.size(),
+            worker_slices_streams[i].size());
+    }
+
+    return worker_slices_streams;
+};
+
+Shape4D<uint32_t> from_tensor_shape(ttnn::Shape const& shape) {
+    constexpr size_t max_rank = 4;
+    TT_FATAL(
```

Comment:
Can the tensor be viewed to an equivalent rank 4 tensor rather than asserting?

Reply:
Makes sense since the collapse may depend on how they want the reduction to happen